File: codex/app/sources/crossref.py
from app.sources.base import SourceContext
from app.utils import safe_title_query


async def fetch_crossref(ctx: SourceContext, doi: str | None, title: str | None) -> dict | None:
    if doi:
        url = f"https://api.crossref.org/works/{doi}"
        data = await ctx.get_json(url)
        message = data.get("message") or {}
        for link in message.get("link", []) or []:
            if link.get("content-type") == "application/pdf" and link.get("URL"):
                return {"pdf_url": link.get("URL"), "metadata": message, "source": "crossref"}
        return None

    if not title:
        return None

    url = f"https://api.crossref.org/works?query.title={safe_title_query(title)}&rows=1"
    data = await ctx.get_json(url)
    items = (data.get("message") or {}).get("items") or []
    if not items:
        return None
    item = items[0]
    for link in item.get("link", []) or []:
        if link.get("content-type") == "application/pdf" and link.get("URL"):
            return {"pdf_url": link.get("URL"), "metadata": item, "source": "crossref"}
    return None

File: claude/src/fetchers/crossref.mjs
import axios from 'axios';
import { isTitleMatch } from '../utils/titleMatch.mjs';

/**
 * Crossref fetcher - searches for papers by title and returns PDF links
 * Uses query.title for more precise matching
 */
export async function fetchFromCrossref(title) {
  try {
    const searchUrl = 'https://api.crossref.org/works';
    const params = {
      'query.title': title,  // More precise title search
      rows: 5,
      select: 'DOI,title,author,published,publisher,link'
    };

    const response = await axios.get(searchUrl, { params, timeout: 15000 });

    if (!response.data.message?.items || response.data.message.items.length === 0) {
      return { success: false, error: 'No results found on Crossref' };
    }

    // Look through results for matching title with PDF
    for (const item of response.data.message.items) {
      const resultTitle = item.title?.[0];
      
      // Validate title match
      if (!isTitleMatch(title, resultTitle)) {
        continue;
      }

      // Check for direct PDF link
      const pdfLink = item.link?.find(l => l['content-type'] === 'application/pdf');
      
      if (pdfLink?.URL) {
        return {
          success: true,
          pdf_url: pdfLink.URL,
          source: 'Crossref',
          doi: item.DOI,
          metadata: {
            title: resultTitle,
            authors: item.author?.map(a => `${a.given || ''} ${a.family || ''}`).join(', '),
            doi: item.DOI,
            publisher: item.publisher
          }
        };
      }

      // If matching title found but no PDF, return DOI for Unpaywall
      if (item.DOI) {
        return {
          success: false,
          error: 'No direct PDF link from Crossref',
          doi: item.DOI,
          metadata: {
            title: resultTitle,
            authors: item.author?.map(a => `${a.given || ''} ${a.family || ''}`).join(', '),
            doi: item.DOI,
            publisher: item.publisher
          }
        };
      }
    }

    return { success: false, error: 'No matching paper found on Crossref' };
  } catch (error) {
    return { success: false, error: `Crossref error: ${error.message}` };
  }
}
